### ISSUE TITLE
Can't call recursive async function inside spawn

### ISSUE THREAD

**Issue by Evian-Zhang** [tokio-rs/tokio#2394]
## Version

rust: 1.42.0
tokio: 0.2.13

## Platform

Linux 5.3.0-46 generic 18.04.1-Ubuntu
## Description
I tried this code:

```rust
async foo() {
    tokio::spawn(async {
        bar().await;
    });
}

async bar() {
    foo().await;
}
```

I expected to see this happen: 

The code should be compiled.

Instead, this happened:

```
error[E0391]: cycle detected when processing `foo`
 --> src/lib.rs:3:1
  |
3 | async fn foo() {
  | ^^^^^^^^^^^^^^
  |
  = note: ...which requires evaluating trait selection obligation `{impl std::future::Future, impl std::future::Future, (), ()}: std::marker::Send`...
note: ...which requires processing `foo::{{opaque}}#0`...
 --> src/lib.rs:3:16
  |
3 | async fn foo() {
  |                ^
  = note: ...which again requires processing `foo`, completing the cycle
note: cycle used when processing `foo::{{closure}}#0`
 --> src/lib.rs:3:16
  |
3 |   async fn foo() {
  |  ________________^
4 | |     tokio::spawn(async {
5 | |         bar().await
6 | |     });
7 | | }
  | |_^

error: aborting due to previous error
```

**Comment by Darksonn** [comment]
This is because for `foo()` to call `tokio::spawn`, it must first create an anonymous variable containing the async block, that it can then pass to `tokio::spawn`. Of course, to create a variable, it must know the type of that variable.

To compute the layout of the type of the async block, it needs to figure out which variables exist inside the block, because those can become fields inside the async block. In this case, the return value of `bar()` (i.e. the `Future` you `.await`) is one such variable.

However in order to store the `Future` returned by `bar()` inside the async block, you must figure out what type `bar()` returns, and this future contains the return value of `foo()` so it can `.await` it, which much then become a field. We've now reached the cycle.

Basically it fails for the same reason that the below fails. Just because you're creating an `Foo { block: AsyncBlock::NotYetStarted() }`, that doesn't mean the type has finite size, due to the _possibility_ of a cycle.
```rust
struct Foo {
    block: AsyncBlock,
}
enum AsyncBlock {
    NotYetStarted(),
    AwaitingBar(Bar),
}
struct Bar {
    foo: Foo,
}
```
For now you can do this:
```rust
async fn foo() {
    foo_spawn();
}
fn foo_spawn() {
    tokio::spawn(async {
        bar().await;
    });
}

async fn bar() {
    foo().await;
}
```
If you want to further pursue this bug, you should post it on the rustc github instead, as it's not related to Tokio.
